esphome/components/image/__init__.py

```diff
@@ -74,6 +77,56 @@ def download_mdi(value):
     return value
 
 
+def get_filename_from_url(url):
+    # Regular expression pattern to match the file name at the end of the URL
+    pattern = r"/([^/]+)$"
+    match = re.search(pattern, url)
+    if match:
+        file_name = match.group(1)
+        file_name = re.sub(r"%20", "", file_name)  # Replace %20 with empty string
+
+        # Remove the file extension
+        file_name = re.sub(r"\.[^.]+$", "", file_name)
+
+        return file_name
+    return None
+
+
+def get_file_type(file_name):
+    # Regular expression pattern to match the file extension
+    pattern = r"\.([^.]+)$"
+    match = re.search(pattern, file_name)
+    if match:
+        file_type = match.group(1)
+        return file_type
+    return None
+
+
+def _compute_local_image_path(value) -> Path:
+    image_id = get_filename_from_url(value[CONF_URL])
+    filetype = get_file_type(value[CONF_URL])
+    base_dir = Path(CORE.config_dir) / ".esphome" / DOMAIN / f"{filetype}"
+    return base_dir / f"{image_id}.{filetype}"
+
+
+def download_image(value):
+    image_id = get_filename_from_url(value[CONF_URL])
+    path = _compute_local_image_path(value)
+    if path.is_file():
+        return value
+    url = value[CONF_URL]
+    _LOGGER.debug("Downloading %s image from %s", image_id, url)
+    try:
+        req = requests.get(url, timeout=MDI_DOWNLOAD_TIMEOUT)
```

Comment:
Since now the timeout is used as well for image download, you might want to rename the constant.
```suggestion
        req = requests.get(url, timeout=IMAGE_DOWNLOAD_TIMEOUT)
```